sdks/python/apache_beam/runners/interactive/pipeline_instrument.py

```diff
@@ -0,0 +1,470 @@
+            key = self._pin.cache_key(input_list[i])
+            if key in self._pin._cached_pcoll_read:
+              input_list[i] = self._pin._cached_pcoll_read[key]
+          transform_node.inputs = tuple(input_list)
+
+    v = ReadCacheWireVisitor(self)
+    self._pipeline.visit(v)
+
+  def _cacheable_inputs(self, transform):
+    inputs = set()
+    for in_pcoll in transform.inputs:
+      if self._cacheable_key(in_pcoll) in self.cacheables():
+        inputs.add(in_pcoll)
+    return inputs
+
+  def _cacheable_key(self, pcoll):
+    """Gets the key a cacheable PCollection is tracked within the instrument."""
+    return cacheable_key(pcoll, self.pcolls_to_pcoll_id(),
+                         self._pcoll_version_map)
+
+  def cache_key(self, pcoll):
+    """Gets the identifier of a cacheable PCollection in cache.
+
+    If the pcoll is not a cacheable, return ''.
+    The key is what the pcoll would use as identifier if it's materialized in
+    cache. It doesn't mean that there would definitely be such cache already.
+    Also, the pcoll can come from the original user defined pipeline object or
+    an equivalent pcoll from a transformed copy of the original pipeline.
+    """
+    cacheable = self.cacheables().get(self._cacheable_key(pcoll), None)
+    if cacheable:
+      return '_'.join((cacheable['var'],
+                       cacheable['version'],
+                       cacheable['pcoll_id'],
+                       cacheable['producer_version']))
+    return ''
+
+  def _debug_appliedptransform(self, transform_node):
+    """Debugs AppliedPTransform at debug level logging.
+
+    Logs structure of an AppliedPTransform instance. Used for testing,
+    debugging and dev purpose.
+    """
+    logging.debug('parent is: %s', transform_node.parent)
+    logging.debug('full label of transform_node: %s',
+                  transform_node.full_label)
+    logging.debug('inputs: ')
+    for inp in transform_node.inputs:
+      logging.debug('  %s', inp)
+    logging.debug('parts: ')
+    for part in transform_node.parts:
+      logging.debug('  %s', part)
+    logging.debug('outputs: ')
+    for output in transform_node.outputs:
+      logging.debug('  %s', output)
+
+  def _debug_pipeline_graph(self):
+    """Debugs the pipeline at debug level logging.
+
+    The pipeline within the class is being instrumented and mutates. This
+    function logs structure of the pipeline exactly when invoked and can be
+    invoked multiple times at different instrumenting stages for testing,
+    debugging and dev purposes.
+    """
+
+    class DebugVisitor(PipelineVisitor):
+      def __init__(self, pin):
+        self._pin = pin
+
+      def enter_composite(self, transform_node):
+        self.visit_transform(transform_node)
+
+      def visit_transform(self, transform_node):
+        self._pin._debug_appliedptransform(transform_node)
+
+    v = DebugVisitor(self)
+    self._pipeline.visit(v)
+
+
+def pin(pipeline, options=None):
```

Comment:
What does `pin` mean?

Reply:
It's short for `pipeline instrument`, also (verb.) `pin`s the `pipeline` at this moment, with original pipeline information snapshotted and caching (interactivity) transforms instrumented once by mutating the pipeline object from the state when it receives the pipeline.

Reply:
Which one it is? Is it the abbreviation or the verb in this context?